src/validations/html/custom.js

```diff
@@ -0,0 +1,144 @@
+import {SAXParser} from 'parse5';
+import voidElements from 'void-elements';
+import Validator from '../Validator';
+
+// An immutable.Record wrapper of parse5.MarkupData.Location, whose line and
+// col are 1-indexed.
+const Location = Record({line: 1, col: 1});  // eslint-disable-line new-cap
+
+function findMismatchedCloseTags(parser) {
+  let openTagStack = List.of();
+  let mismatchedTagErrorsByLocation = Map.of();
+  let unopenedTagLocations = Set.of();
+  let unopenedTagQueuesByName = Map.of();
+
+  parser.on('startTag', (name, attrs, selfClosing, location) => {
+    if (!(selfClosing || name in voidElements)) {
+      openTagStack = openTagStack.push({
+        name,
+        location: new Location(location),
+      });
+    }
+  });
+
+  parser.on('endTag', (name, locationObject) => {
+    const location = new Location(locationObject);
+
+    const openIndex = openTagStack.findLastIndex(
+      openTag => openTag.name === name);
+    const opened = openIndex >= 0;
+
+    // Assume any open tags after the matching open tag are mismatched. If the
+    // source is valid HTML, then no other open tags should exist. If no
+    // matching open tag exists, then all open tags are mismatched.
+    const mismatchedTags =
+      opened ? openTagStack.slice(openIndex + 1) : openTagStack;
+    mismatchedTagErrorsByLocation = mismatchedTagErrorsByLocation.set(
+      location, {name, mismatchedTags});
+
+    if (opened) {
+      openTagStack = openTagStack.slice(0, openIndex);
+    } else {
+      unopenedTagLocations = unopenedTagLocations.add(location);
+      unopenedTagQueuesByName =
+        unopenedTagQueuesByName.update(
+          name, List.of(), tags => tags.push({name, location}));
+
+      // If no matching open tag exists, then assume later close tags could
+      // match some open tag instead of emptying openTagStack.
+    }
+  });
+
+  return function* () {
+    // Suggest misplaced-close-tag if a spare unopened tag exists. If no spare
+    // exists:
+    //
+    // A mismatched close tag that has been opened means the case
+    //
+    // <div>
+    //   <span>
+    // </div>
+    //
+    // so suggest unclosed-tag-by-mismatched-close-tag.
+    //
+    // A mismatched close tag that has yet to be opened means the case
+    //
+    //   <span>
+    //   </span>
+    // </div>
+    //
+    // so suggest unexpected-close-tag.
+    for (const [location, {name, mismatchedTags}]
+         of mismatchedTagErrorsByLocation) {
+      const opened = !unopenedTagLocations.has(location);
+      for (const mismatchedTag of mismatchedTags) {
+        const spareUnopenedTag =
+          unopenedTagQueuesByName.get(mismatchedTag.name, List.of()).first();
+        if (spareUnopenedTag) {
+          unopenedTagQueuesByName = unopenedTagQueuesByName.update(
+            spareUnopenedTag.name, tags => tags.shift());
+
+          yield {
+            row: location.get('line') - 1,
+            column: location.get('col') - 1,
```

Comment:
I would make a slightly stronger claim that the location of this error should be the location of the `spareUnopenedTag` rather than the location of the mismatched closing tag, e.g.:

```html
<!DOCTYPE html>
<html>
    <head>
        <title>Page Title</title>
    </head>
    <body>
        <div>
            <p>
        </div>
        </p> <!-- fundamentally feels like the error is on this line -->
    </body>
</html>
```

I would change the message to e.g. `This closing </p> tag should go before the </div> on line 9`